### PR TITLE
Improve image classification example

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR improves the image classification example script, by:
- replacing the `nateraw/imagefolder` hack with [`ImageFolder`](https://huggingface.co/docs/datasets/v2.0.0/en/image_process#imagefolder)
- improving the README, explaining how to easily train on custom data.

### CODE REVIEW

File: examples/pytorch/image-classification/run_image_classification.py [hunk 1]
```diff
@@ -201,25 +201,44 @@ def main():
...
+            )
+        if data_args.validation_dir is not None:
+            dataset_dict["validation"] = load_dataset(
+                "imagefolder",
+                data_dir=data_args.validation_dir,
+                cache_dir=model_args.cache_dir,
+                task="image-classification",
+                split="validation",
+            )
+        dataset = DatasetDict(dataset_dict)
```
Comment (NielsRogge) [inline examples/pytorch/image-classification/run_image_classification.py:230]: cc @nateraw @mariosasko , this is how I'm loading local folders now.

File: examples/pytorch/image-classification/README.md [hunk 2]
```diff
@@ -87,24 +77,57 @@ root/cat/nsdf3.png
...
+from datasets import load_dataset
+
+# example 1: local folder
+dataset = load_dataset("imagefolder", data_dir="path_to_your_folder")
+
+# example 2: local files (suppoted formats are tar, gzip, zip, xz, rar, zstd)
+dataset = load_dataset("imagefolder", data_files="path_to_zip_file")
+
+# example 3: remote files (suppoted formats are tar, gzip, zip, xz, rar, zstd)
+dataset = load_dataset("imagefolder", data_files="https://download.microsoft.com/download/3/E/1/3E1C3F21-ECDB-4869-8368-6DEBA77B919F/kagglecatsanddogs_3367a.zip"
```
Comment (sgugger) [inline examples/pytorch/image-classification/README.md:111]:
```suggestion
dataset = load_dataset("imagefolder", data_files="https://download.microsoft.com/download/3/E/1/3E1C3F21-ECDB-4869-8368-6DEBA77B919F/kagglecatsanddogs_3367a.zip")
```

**Review by sgugger — APPROVED** [review]
Very cool, thanks a lot for making this script better!

File: examples/pytorch/image-classification/run_image_classification.py (hunk 1 above)
Comment (mariosasko) [reply examples/pytorch/image-classification/run_image_classification.py:230]: ImageFolder + `data_dir` collects everything in a `train` split, so this part is problematic:
```python
 dataset_dict["validation"] = load_dataset(
                "imagefolder",
                data_dir=data_args.validation_dir,
                cache_dir=model_args.cache_dir,
                task="image-classification",
                split="validation", <-
            )
```

We can rewrite this part as follows to make it cleaner and more robust (for instance, this approach doesn't fail if some labels are not present in the validation `data_dir`):
```python
data_files = {}
if data_args.train_dir is not None:
    data_files["train"] = os.path.join(data_args.train_dir, "**")
if data_args.validation_dir is not None:
    data_files["validation"] = os.path.join(data_args.validation_dir, "**")
dataset = load_dataset(
    "imagefolder",
     data_files=data_files,
     cache_dir=model_args.cache_dir,
     task="image-classification",
)
```

**Pull request merged**